src/tokenizer/regex_tokenizer.rs

```diff
@@ -0,0 +1,149 @@
+use regex::Regex;
+use crate::TantivyError;
+
+use super::{BoxTokenStream, Token, TokenStream, Tokenizer};
+
+/// Tokenize the text by using a regex pattern to split.
+/// Each match of the regexp emits an distinct tokens, empty tokens will not be emitted.
+///
+///
+/// Example: `` 'aaa' bbb 'ccc' 'ddd' `` with the pattern `` '(\w*)' `` will be tokenized as followed:
+///
+/// | Term     | aaa  | ccc    | ddd   |
+/// |----------|------|--------|-------|
+/// | Position | 1    | 2      | 3     |
+/// | Offsets  |0,5   | 10,15  | 16,21 |
+///
+///
+/// # Example
+///
+/// ```rust
+/// use tantivy::tokenizer::*;
+///
+/// let tokenizer = RegexTokenizer::new(r"'(\w*)'");
+/// let mut stream = tokenizer.token_stream("'aaa' bbb 'ccc' 'ddd'");
+/// {
+///     let token = stream.next().unwrap();
+///     assert_eq!(token.text, "'aaa'");
+///     assert_eq!(token.offset_from, 0);
+///     assert_eq!(token.offset_to, 5);
+/// }
+/// {
+///   let token = stream.next().unwrap();
+///     assert_eq!(token.text, "'ccc'");
+///     assert_eq!(token.offset_from, 10);
+///     assert_eq!(token.offset_to, 15);
+/// }
+/// {
+///   let token = stream.next().unwrap();
+///     assert_eq!(token.text, "'ddd'");
+///     assert_eq!(token.offset_from, 16);
+///     assert_eq!(token.offset_to, 21);
+/// }
+/// assert!(stream.next().is_none());
+/// ```
+
+#[derive(Clone)]
+pub struct RegexTokenizer {
+    regex: Regex
+}
+
+impl RegexTokenizer {
+    pub fn new(regex_pattern: &str) -> crate::Result<RegexTokenizer> {
+        return Regex::new(regex_pattern)
+            .map_err(|_| TantivyError::InvalidArgument(regex_pattern.to_string()))
+            .map(|p| RegexTokenizer { regex: p });
+    }
+}
+
+impl Tokenizer for RegexTokenizer {
+    fn token_stream<'a>(&self, text: &'a str) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(RegexTokenStream {
+            regex: self.regex.clone(),
```

Comment:
The regex currently is cloned here, this is inefficient. However, the regex should be only compiled once per Regex Tokenizer, so I think it needs to be passed down. I think this would need another lifetime annotation for the regex, which would require to modify BoxTokenStream. I s there are better way to do this ?

Reply:
You could store `Arc<Regex>` everywhere and clone that. This is done by e.g. the `StopWordFilter` and `SplitCompoundWords` filters.

But I think this is a non-obvious trade-off: `Regex` is internally an `Arc<..>` of a read-only part itself and the second part is a pool of cache memory for the various search algorithms. So cloning the `Regex` itself might actually fit this use case pretty well, c.f. https://github.com/rust-lang/regex/blob/master/PERFORMANCE.md#using-a-regex-from-multiple-threads